### PR TITLE
[feature] Supporting half precision floating data type (fp16).

### PR DESCRIPTION
**Pull request by yzh119**
## Description
Refactor the code to support fp16, compatible with torch.cuda.amp according to experiments.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [x] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: src/array/cpu/sddmm.cc [hunk 1]
```diff
@@ -41,9 +41,22 @@ namespace aten {
     }                                                                   \
   } while (0)
 
+#define SWITCH_BITS(bits, DType, ...)                           \
+  do {                                                          \
+    if ((bits) == 16 || (bits) == 32) {                         \
+      typedef float DType;                                      \
```
Comment (BarclayII) [inline src/array/cpu/sddmm.cc:47]: Do we have comments or documentation saying the CPU FP16 is done with FP32?

Comment (yzh119) [reply src/array/cpu/sddmm.cc:47]: No and I'll introduce this behavior in FAQ if you think this PR is ready to be merged.

File: CMakeLists.txt [hunk 2]
```diff
@@ -25,6 +25,7 @@ endif()
 dgl_option(USE_CUDA "Build with CUDA" OFF)
 dgl_option(USE_OPENMP "Build with OpenMP" ON)
 dgl_option(USE_AVX "Build with AVX optimization" ON)
+dgl_option(USE_AMP "Build with fp16 support to enable AMP(automatic mixed precision)" OFF)
```
Comment (yzh119) [inline CMakeLists.txt:28]: Disabled by default because we only support CUDA ARCH>=530, while our binary supports sm35 and sm50.
User could manually turn this flag on if they would like to train GNN with AMP.

Comment (yzh119) [reply CMakeLists.txt:28]: @BarclayII FYI.

**Review by jermainewang — APPROVED** [review]
Overall very good. Some comments.

File: src/array/cpu/sddmm.cc [hunk 3]
```diff
@@ -41,9 +41,22 @@ namespace aten {
...
+      { __VA_ARGS__ }                                           \
+    } else {                                                    \
+      LOG(FATAL) << "Data type not renogized with bits " << bits; \
+    }                                                           \
+  } while (0)
+
 
 /*! \brief Generalized SDDMM on Csr format. */
-template <int XPU, typename IdType, typename DType>
+template <int XPU, typename IdType, int bits>
```
[the remark this answers is not in the document]
Comment (yzh119) [reply src/array/cpu/sddmm.cc:59]: Currently yes.

File: src/array/cuda/spmm.cu [hunk 4]
```diff
@@ -14,6 +14,140 @@ namespace dgl {
...
+namespace {
+
+/*! \brief Call cuBLAS geam API for transpose operation for float and double. */
+template <typename DType>
+cublasStatus_t Xgeam(cublasHandle_t handle, cublasOperation_t transa,
+    cublasOperation_t transb, int m, int n,
+    const DType* alpha, const DType* A, int lda,
+    const DType* beta, const DType* B, int ldb,
+    DType* C, int ldc) {
+  LOG(INFO) << "Not supported dtype";
```
[the remark this answers is not in the document]
Comment (yzh119) [reply src/array/cuda/spmm.cu:26]: This should be unreachable.

File: src/array/cuda/utils.h [hunk 5]
```diff
@@ -18,6 +18,37 @@ namespace cuda {
 #define CUDA_MAX_NUM_BLOCKS_Z 0xFFFF
 #define CUDA_MAX_NUM_THREADS 1024
 
+#ifdef USE_FP16
+#define SWITCH_BITS(bits, DType, ...)                           \
```
[the remark this answers is not in the document]
Comment (yzh119) [reply src/array/cuda/utils.h:22]: I have removed that.

File: src/array/kernel.cc [hunk 6]
```diff
@@ -226,6 +235,12 @@ DGL_REGISTER_GLOBAL("sparse._CAPI_DGLKernelBwdSegmentCmp")
     BackwardSegmentCmpDispatch(feat, arg, out);
   });
 
+DGL_REGISTER_GLOBAL("sparse._CAPI_DGLKernelGetEdgeMapping")
+.set_body([](DGLArgs args, DGLRetValue *rv) {
+    HeteroGraphRef graph = args[0];
+    *rv = GetEdgeMapping(graph);
+  });
```
[the remark this answers is not in the document]
Comment (yzh119) [reply src/array/kernel.cc:242]: No, I used it to measure the performance difference between:
1. use edge mapping inside kernels (incurs extra and irregular global memory access)
2. shuffle the feature according to edge mapping and feed it to kernel.

Sometimes 1 is better and sometimes 2 is better, it's also a tunable parameter.

File: docs/source/mixed_precision.rst [hunk 7]
```diff
@@ -0,0 +1,78 @@
+Mixed Precision Training
```
[the remark this answers is not in the document]
Comment (yzh119) [reply docs/source/mixed_precision.rst]: DONE

File: python/dgl/backend/pytorch/sparse.py [hunk 8]
```diff
@@ -1,7 +1,31 @@
...
+    """PyTorch natively supports automatic mixed precision in DGL 1.6, we redefine
+    the custom_fwd and custom_bwd function to be compatible with DGL 1.5.
+    """
+    def custom_fwd(**kwargs):
+        def custom_fwd_inner(fwd):
+            @functools.wraps(fwd)
+            def decorate_fwd(*args, **kwargs):
+                if fwd is None:
+                    return functools.partial(custom_fwd, )
+                return fwd(*args, **kwargs)
```
[the remark this answers is not in the document]
Comment (yzh119) [reply python/dgl/backend/pytorch/sparse.py:17]: Yes the outer kwargs is use nowhere in this case.

File: src/array/cuda/functor.cuh [hunk 9]
```diff
@@ -246,10 +275,21 @@ struct Min {
...
+          typename DType,
+          bool atomic=false>
+struct Min : _Min<Idx, DType, atomic> { };
+
+#ifdef USE_FP16
+template <typename Idx,
+          bool atomic>
+struct Min<Idx, half, atomic> : _Min<Idx, half, atomic> {
+  static constexpr __host__ __device__ __forceinline__ half zero() {
+    return __float2half_rn(6.550400e+04f);
```
[the remark this answers is not in the document]
Comment (yzh119) [reply src/array/cuda/functor.cuh:289]: No this function has a long history in CUDA.

File: src/array/cuda/segment_reduce.cu [hunk 10]
```diff
@@ -6,77 +6,105 @@
...
-  } else if (op == "max") {
-    cuda::SegmentReduce<IdType, DType, cuda::reduce::Max<IdType, DType>>(
-        feat, offsets, out, arg);
-  } else if (op == "min") {
-    cuda::SegmentReduce<IdType, DType, cuda::reduce::Min<IdType, DType>>(
-        feat, offsets, out, arg);
-  } else {
-    LOG(FATAL) << "Not implemented";
-  }
+  SWITCH_BITS(bits, DType, {
```
[the remark this answers is not in the document]
Comment (yzh119) [reply src/array/cuda/segment_reduce.cu:24]: `SWITCH_BITS` switch nbits to a data type.

File: src/array/cuda/utils.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (yzh119) [reply src/array/cuda/utils.h:22]: `ATEN_FLOAT_BITS_SWITCH` acts differently, it switches DLDataType (defined in dlpack) to nbits.
